[베르] 코틀린 자동차 경주 제출합니다.

SeongHoonC opened the pull request:
페어 프로그래밍 하면서 많이 배우고 고민할 수 있었습니다.
이번 미션에 질문 있습니다.

다음 방법 중 어떤 것이 더 좋을지 고민입니다.
1. Car 객체를 CarsDTO에서 생성하고 Cars List를 만들어 넘겨주기
2. Controller 에서 자동차의 names 를 받아 Car 객체를 만들고 Cars 리스트 생성하기
3. 새로운 클래스 만들어 처리하기
4. 그 외

ippnsj committed:
docs: 기능 목록 작성

ippnsj committed:
feat: 스켈레톤 코드 작성

ippnsj committed:
docs: 기능 목록 수정

자동차 최대 갯수 20대로 변경
자동차 이름 중복 체크 기능 추가

ippnsj committed:
feat: 경주할 자동차 이름을 입력받는 기능 구현

ippnsj committed:
feat: 자동차 이름 유효성 검사 기능 구현

자동차 개수는 2대 이상 20대 이하여야 한다
자동차 이름은 공백 없는 영문 소문자로만 이루어져야 한다
자동차 이름은 1글자 이상 5글자 이하여야 한다
자동차 이름은 중복되지 않아야 한다

ippnsj committed:
feat: 경주 시도 횟수를 입력받는 기능 구현

ippnsj committed:
feat: 경주 시도 횟수 유효성 검사 기능 구현

시도 횟수는 숫자여야 한다
시도 횟수는 1이상 10이하여야 한다

ippnsj committed:
feat: 0에서 9 숫자 랜덤 생성 기능 구현

ippnsj committed:
feat: 자동차 이동 기능 구현

0 ~ 3이면 stop
4 ~ 9면 go

ippnsj committed:
feat: 거리가 가장 큰 최종 우승자를 찾는 기능 구현

ippnsj committed:
test: 자동차 이동 테스트 추가

ippnsj committed:
feat: 실행 결과 출력 기능 구현

ippnsj committed:
feat: 최종 우승자 출력 기능 구현

ippnsj committed:
refactor: 도메인에서의 뷰 의존성 제거

Car 클래스 내부에서 getRaceResult() 제거
race()의 return 값 RaceResultDTO -> CarMetadataDTO 변경

ippnsj committed:
feat: 경주 전 자동차 입력 및 생성

ippnsj committed:
feat: 경주 전 시도 횟수 설정

ippnsj committed:
feat: 시도 횟수만큼 자동차 경주

ippnsj committed:
feat: 최종 우승자 찾고 출력

ippnsj committed:
feat: 자동차 경주 구현

ippnsj committed:
refactor: 인덴트 요구 사항에 맞게 코드 분리

ippnsj committed:
docs: 기능 목록 수정

맞춤법 및 줄 정리

laco-dev commented on src/main/kotlin/Application.kt:
```diff
@@ -0,0 +1,2 @@
+fun main() {
```
> feat: 스켈레톤 코드 작성

기능을 구현하는 과정에서 어떻게 구현을 해 나갈지 모르기 때문에 이러한 커밋 단위는 좋지 않은 방법입니다.
실제로 구현할때 코드가 변경이 될 텐데, 왜 이렇게 만들어지기 시작했는지 작성자 본인 외에는 아무도 알기 어려울 것입니다.

귀찮음을 피하는 것은 커밋으로 해결하는 것이 아닌, 다른 방법을 모색해야 합니다.

SeongHoonC replied:
넵! 편의 보다는 기능 단위로 커밋 하겠습니다!

laco-dev commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -1,3 +1,19 @@
 package view
 
-class InputView
+import dto.CarNamesDTO
```
입력 기능을 먼저 구현 해 주셨네요. 👍🏻

앞으로 여러가지 도메인을 구현하는 과정에서의, 한 가지 팁이라면 도메인 로직을 먼저 구현해보면 어떨까? 라는 생각을 가지면 좋습니다.

가령, 자동차 이름을 입력받는 기능을 구현한 뒤 자동차 이름이라는 CarName 모델이 생기게 된다면,
InputView의 구현이 따라서 변경이 되야 합니다.
즉 코드를 두번 이상 수정해야 한다는 의미가 될 수 있어요.

꼭 지켜야 하는 규칙은 아니지만, 기회가 된다면 한번 경험해보시면 좋습니다.

SeongHoonC replied:
넵! 다음 미션은 도메인부터 구현해 보도록 하겠습니다!

laco-dev commented on src/main/kotlin/dto/CarNamesDTO.kt:
```diff
@@ -0,0 +1,43 @@
+package dto
+
+class CarNamesDTO(input: String) {
```
> feat: 자동차 이름 유효성 검사 기능 구현
자동차 개수는 2대 이상 20대 이하여야 한다
자동차 이름은 공백 없는 영문 소문자로만 이루어져야 한다
자동차 이름은 1글자 이상 5글자 이하여야 한다
자동차 이름은 중복되지 않아야 한다

DTO의 역할은 무엇일까요?
DTO가 필요한 곳마다 이와 같은 검증 로직이 필요하다면, 모든 계층마다 매번 같은 코드를 작성해야 하지는 않을까요?

도메인 모델과 DTO에 대한 차이점을 생각 해 보시면 좋을 것 같아요.

SeongHoonC replied:
DTO 는 layer 간 data를 전달하는 바구니로 로직을 포함해서는 안되는 것을 알게 되었습니다. 

검증 로직을 도메인 내부에 만들어야 할 것 같습니다!

laco-dev commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -0,0 +1,73 @@
+    private val inputView: InputView,
+    private val outputView: OutputView
+) {
+    private val cars = mutableListOf<Car>()
+    private val judgement = Judgement(cars)
+    private var time = 0
```
이 값들이 멤버 변수로 만들어질 필요가 있을까요?
멤버 변수로 만들어지면서 어떤 장단점이 생기게 되었을까요?

함수에서 필요한 값을 전달하는 또 다른 방법은 무엇일까요?

SeongHoonC replied:
- 장점 
사용할 주요 객체를 미리 명시할 수 있고 같은 클래스 내에서 인자로 넘겨주지 않아도 되는 편의성을 제공합니다.

- 단점 
클래스가 인스턴스화되면 초기화되어 불필요하게 메모리를 차지하고 있습니다. 만약 같은 게임을 여러번 하면 재사용이 힘들어 전체 구조를 바꿔야할 수 있습니다.

- 생각
개발자에게 편의성을 제공하는게 효율적인 코드는 아니기 때문에 멤버 변수로 만들지 않고 필요할 때 생성하는게 좋을 것 같습니다.

laco-dev replied:
잘 정리 해주셨네요. 👍🏻 

작성해주신대로, 변수가 필요한 범위를 가장 좁은 곳부터 시작해서 확장 해 나가는 방법이 좋겠습니다.
정말 멤버 변수로 필요한지는 생각해보면 금방 답을 찾을 수 있습니다.

laco-dev commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -0,0 +1,73 @@
+        var completeSetUpRaceTime = false
+        while (!completeSetUpCar) {
+            completeSetUpCar = setUpCars()
+        }
```
반복문을 포함해서 setupCars에게 역할을 맡길 수 있지 않을까요?

```kotlin
val cars = getReadyCars()
val raceTime = getRaceTime()
...
```

laco-dev commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -0,0 +1,73 @@
+
+    private fun race() {
+        outputView.printExecutionResult()
+        for (t in 0 until time) {
+            raceOneTime()
+            outputView.printInterval()
+        }
```
코틀린의 repeat 함수를 활용할 수 있어요!
```kotlin
repeat(time) { ... }
```

SeongHoonC replied:
훨씬 간결해지네요. 감사합니다!

laco-dev commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private fun raceOneTime() {
+        cars.map { car ->
+            outputView.printRaceResult(car.race())
+        }
```
map의 올바른 사용 방법이 맞을까요?

SeongHoonC replied:
map 은 뒤에서 정의된 리스트를 반환하는데 여기선 forEach로 바꾸는게 적합할 것 같습니다!

laco-dev commented on src/main/kotlin/data/ComparisonResult.kt:
```diff
@@ -0,0 +1,5 @@
+package data
+
+enum class ComparisonResult {
+    WIN, DRAW, LOSE, NONE
+}
```
비즈니스 개념으로 볼 때, NONE이라는 값은 무엇을 의미하는지 예측이 될까요?
승, 패, 무승부를 제외하고는 어떤 상황에 NONE이 발생할 수 있을지 알기 어렵지는 않을까요?

SeongHoonC replied:
상태가 전달되지 못했을 경우를 생각했는데 그 경우 null 을 전달하고 null 체크 하는 것이 더 kotlin 스러울 것 같습니다.

NONE은 빼는 것이 나을 것 같네요!

laco-dev commented on src/main/kotlin/data/ComparisonResult.kt:
```diff
@@ -0,0 +1,5 @@
+package data
```
생각하시는 data 패키지의 역할은 무엇이라고 생각하시나요?
이 클래스가 domain에 위치할 수 없는 이유가 있을까요?

SeongHoonC replied:
data 는 서버에서 값을 받아오거나 가공하기 위한 패키지로 생각합니다.

domain 에 위치하지 못하는 이유는 없습니다.

laco-dev commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,43 @@
+            difference < 0 -> ComparisonResult.LOSE
+            else -> ComparisonResult.DRAW
+        }
+
+        return CarMetadataDTO(comparisonResult, distance, name)
+    }
+
+    private fun checkGo(number: Int): Boolean {
+        if (number in RANGE_LOWER_INCLUSIVE..RANGE_UPPER_INCLUSIVE)
+            return true
+        return false
+    }
```
> 값이 4 이상일 경우이다.

자동차가 이동하는 조건은 4 이상으로만 볼 수 있지 않을까요?
0 부터 9 사이의 값이 발행되는 것은 랜덤의 관심사로 볼 수 있을 것 같아요.

SeongHoonC replied:
4 이상만 판단하도록 변경하겠습니다!

laco-dev commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,43 @@
+        return CarMetadataDTO(ComparisonResult.NONE, distance, name)
+    }
+
+    fun compare(winnerDistance: Int): CarMetadataDTO {
```
비교하는 과정에서 winner라는 것은 상대적이지 않을까요?
이기거나 지거나 하는 결과가 나올 수 있어서 `other` 같은 이름도 괜찮을 것 같아요.

SeongHoonC replied:
이전에 승리한 값을 그대로 저장하려고 이렇게 지었는데 말씀대로 상대적인 것 같습니다.
이름 추천해 주셔서 감사합니다!

laco-dev commented on src/main/kotlin/domain/Judgement.kt:
```diff
@@ -0,0 +1,36 @@
+import dto.WinnersDTO
+
+class Judgement(private val cars: List<Car>) {
+    private var winnerDistance = 0
+    private val winners = mutableListOf<String>()
+
+    fun findWinners(): WinnersDTO {
+        cars.map { car -> updateWinners(car.compare(winnerDistance)) }
+        return WinnersDTO(winners)
+    }
```
이 객체에 대해서도, 멤버 변수로 값을 관리하면 어떤 장점이 있을지 생각해보면 좋을 것 같습니다.

하고싶은 목적은 "우승자 목록을 반환한다" 에 기준을 두고,
조금 더 간단하게 그리고 보기 쉽게 개선할 수는 없을까? 를 생각 해 볼 수 있겠습니다.

SeongHoonC replied:
Car의 distance 중 max 값을 저장하고 max 값과 같은 distance를 가진 Car들의 이름을 리스트로 변환하게 변경하겠습니다.

laco-dev commented on src/main/kotlin/dto/CarMetadataDTO.kt:
```diff
@@ -0,0 +1,21 @@
+
+import data.ComparisonResult
+
+class CarMetadataDTO(
```
CarMeta라는 이름이 무슨 역할을 하는지 알 수 있을까요?

도메인 개념과 비교해 본다면, "경주 결과" 라는 이름과 역할을 가진다고 볼 수 있지 않을까요?

SeongHoonC replied:
경주 결과에 따른 차 정보를 담는데 이름에서 역할을 알 수 있도록 바꾸겠습니다!

laco-dev commented on src/main/kotlin/dto/CarMetadataDTO.kt:
```diff
@@ -0,0 +1,21 @@
+    private val distance: Int,
+    private val name: String
+) {
+    fun getComparisonResult(): ComparisonResult {
+        return comparisonResult
+    }
+
+    fun getDistance(): Int {
+        return distance
+    }
+
+    fun getName(): String {
+        return name
+    }
```
코틀린은 기본적으로 프로퍼티에 대해 getter(), setter()를 만들어 주기 때문에 단순 프로퍼티로 사용된다면 별도의 getter() 함수는 만들지 않아도 괜찮습니다!

SeongHoonC replied:
getter 함수를 없애고 data 클래스로 변경하겠습니다!

laco-dev commented on src/main/kotlin/dto/CarsDTO.kt:
```diff
@@ -0,0 +1,54 @@
+import domain.Car
+import domain.RandomNumberGenerator
+
+class CarsDTO(input: String) {
```
앞서 말씀드린 대로, CarsDTO 라는 DTO가 어떤 역할을 해야 하는지, 고민이 필요합니다.

laco-dev commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -0,0 +1,23 @@
+
+class CarTest {
+    @Test
+    fun `자동차 이동 테스트`() {
```
테스트 코드는 프로그램의 명세서라고 부르기도 합니다.

즉 테스트 이름으로 하여금 어떤 동작을 하는지 예측이 가능하도록 작성을 해야 합니다.
"자동차 이동 테스트"가 무엇을 하는 기능인지, 실패한다면 왜 실패하는지 성공한다면 왜 성공하는지 기능 단위로 명확하게 나타낼 필요가 있겠습니다.

예를 들어 이렇게 작성할 수도 있겠습니다.
```
"자동차는 4 이상의 값이면 한 칸 앞으로 전진한다"
"자동차는 4 미만의 값이면 정지한다"
```

laco-dev commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,43 @@
+            else -> ComparisonResult.DRAW
+        }
+
+        return CarMetadataDTO(comparisonResult, distance, name)
```
DTO는 계층간의 의존성 없이 데이터를 전달하기 위해 사용을 합니다.
그렇다면 도메인 모델이 직접적으로 DTO를 아는 것이 올바른 방법일까요?

laco-dev reviewed (changes requested):
안녕하세요.
자동차 경주 미션과 함께 하게 되었습니다. 😄 

여러가지 복잡한 개념과 모델 분리 등 다양한 방법을 사용 해 주셔서 재미있게 리뷰할 수 있었습니다.

하지만, 각 역할을 명확히 하지 않은채 다소 무분별하게 늘어난 듯한 느낌도 있습니다.
과제를 진행하면서 모든 계층에 DTO를 만들어야 한다 등의 제약 사항은 없지만, DTO와 도메인 모델, 그리고 UI 사이에서의 관계를 충분히 이해한 다음 사용이 필요하겠습니다.

관련해서 몇가지 의견을 남겨드렸으니 충분히 고민 해 보시고 확인 부탁드립니다. 💪🏻

SeongHoonC committed:
refactor: Car 내부에서 이름 검증

Car 에서 글자 수 및 글자 검증

SeongHoonC committed:
refactor: CarsDTO 삭제 및 자동차들을 생성하는 공장 추가

CarsDTO 는 불필요하므로 삭제
자동차들을 생성하는 공장 CarsFactory 정의
자동차 공장에서 자동차의 개수 및 이름 중복 검증

SeongHoonC committed:
refactor: 자동차 이름 입력 시 리스트화해서 전달

CarsDTO 를 사용하지 않고 이름 리스트를 반환

SeongHoonC committed:
refactor: 자동차 공장으로 자동차들을 만들고 인자로 전달

kotlin repeat, forEach 적용

SeongHoonC committed:
refactor: 클래스 이름을 역할에 맞게 변경

CarMetaDataDTO 를 RaceResultDTO 이름 변경

SeongHoonC committed:
refactor: RaceResultDTO 를 data Class 로 변한

SeongHoonC committed:
refactor: WinnersDTO 를 data Class 로 변한

SeongHoonC committed:
refactor: 자동차 이동 판단 범위를 4 이상으로 변경

SeongHoonC committed:
refactor: ComparisonResult 클래스 패키지 변경

data 에서 domain 으로 변경
data 패키지 삭제

SeongHoonC committed:
refactor: 최종 우승자를 찾는 방법 쉽게 변경

car 에서 distance 를 받고 maxDistance 를 가진 Car 의 name 을 List로 만들어 반환
comparisonResult, winnersDTO 는 필요 없으므로 삭제

laco-dev commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -36,8 +40,22 @@ class Car(
 
+    private fun validateCarName() {
+        require(name.length in NAME_LOWER_LENGTH..NAME_UPPER_LENGTH) { ERROR_NAME_LENGTH }
+        name.forEach { require(it.code in NAME_LOWER_LETTER..NAME_UPPER_LETTER) { ERROR_NAME } }
```
캐릭터를 검사하는 방법을 직접 계산하는 대신, 코틀린의 함수를 사용해 볼 수 있겠습니다.

```kotlin
char.isDigit()
char.isLetter()
...
```

등 여러가지 함수를 지원하고 있는데요. 
어떻게 동작하는지 학습 테스트를 실천 해보셔도 좋습니다

laco-dev commented on src/test/kotlin/domain/CarTest.kt:
```diff
@@ -1,18 +1,44 @@
 
+    @ParameterizedTest
+    @ValueSource(strings = ["pobi", "woni", "jun"])
+    fun `자동차 이름에 오류가 없으면 에러가 발생하지 않는다`(input: String) {
```
오류? 라는 내용도 읽는 사람에 따라 어떻게 해야 오류가 없는것인지 생각할 수있어요.
테스트 코드는 단순히 사양을 그대로 표현하는 것이 충분할 때가 많습니다.

ex) "자동차 이름은 1글자에서 5글자 사이의 문자로 되어 있다"

laco-dev commented on src/main/kotlin/domain/CarsFactory.kt:
```diff
@@ -0,0 +1,26 @@
+
+    companion object {
+        const val ERROR_CAR_COUNT = "자동차 개수는 2대 이상 20대 이하여야 합니다."
+        const val ERROR_CAR_UNIQUENESS = "자동차 이름은 중복되면 안 됩니다."
```
자동차 이름이 중복될 수 없는 내용도 마찬가지로 "자동차들" 이라는 도메인 모델로 사용이 될 수 있는데요.

이 개념은 점차 배워나갈 일급 컬렉션이나 객체 포장등을 학습하는 과정에서 자연스럽게 배우실 수 있을 거에요.
지금은 이러한게 있다 정도로만 참고 해 주시면 되겠습니다.

```kotlin
class Cars(...) {
   init {
      require(자동차 이름은 중복할 수 없음)
   }
}
```

laco-dev commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -1,23 +1,16 @@
 
 class InputView {
-    fun readCars(): CarsDTO {
+    fun readCars(): List<String> {
```
함수의 이름은 readCars()보다 readCarNames()가 더 어울릴 것 같네요!

laco-dev commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -1,44 +1,32 @@
-            outputView.printError(e.message ?: "")
-            false
+    private fun setUpCars(): List<Car> {
+        while (true) {
+            try {
+                return CarsFactory(inputView.readCars()).makeCars(RandomNumberGenerator())
+            } catch (e: IllegalArgumentException) {
+                outputView.printError(e.message ?: "")
+            }
```
정상적으로 값이 들어올 때 까지 반복해서 호출하는 것으로 보이네요.

1. runCatching을 사용 해 볼 수 있습니다.
2. 강제로 while 반복문을 사용하는 것 보다는, 재귀 호출로 만들어 보면 어떨까요?

laco-dev commented on src/main/kotlin/controller/Controller.kt:
```diff
@@ -1,44 +1,32 @@
-        race()
-        announceWinners()
+        val cars = setUpCars()
+        setUpRaceTime()
```
setup이라는 함수 이름이 무언가를 반환하는 것과는 잘 매칭이 되지 않아요.
(내부적으로 무언가 상태를 변경하는 것 같은 느낌)

반환 타입에 맞춰 함수 이름을 적절하게 잡아보면 어떨까요?
get(), find(), create() 등등 여러가지 이름을 붙일 수 있습니다

laco-dev commented on src/main/kotlin/dto/RaceResultDTO.kt:
```diff
@@ -0,0 +1,15 @@
+
+    fun getDistance() = distance
+
+    fun getName() = name
```
코틀린에서는 프로퍼티의 접근자를 이용해서 getter(), setter()를 변경할 수 있습니다.
별도의 동작이 없다면 프로퍼티를 활용 해보세요!


https://kotlinlang.org/docs/properties.html#backing-fields

laco-dev commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -0,0 +1,50 @@
+    private var distance: Int = 0
+) {
+
+    init {
+        validateCarName()
+    }
```
마이너한 의견입니다
: 코드를 위에서부터 자연스럽게 읽어 나가기 때문에 함수 순서를 맞춰 주시면 좋아요!

```kotlin
init {
   validateCarName()
}

private fun validateCarName(...)
```

laco-dev commented on src/main/kotlin/dto/RaceTimeDTO.kt:
```diff
@@ -0,0 +1,28 @@
+        require(time in 1..10) { ERROR_NOT_IN_RANGE }
+    }
+
+    companion object {
+        const val ERROR_NOT_INTEGER = "시도 횟수는 숫자여야 합니다."
+        const val ERROR_NOT_IN_RANGE = "시도 횟수는 1이상 10 이하여야 합니다."
+    }
```
다시 드려보는 질문입니다.
1. DTO의 역할은 무엇일까요?
2. DTO가 어떠한 로직을 수행하는 것이 맞을까요?
3. 도메인 모델이 직접적으로 DTO를 만드는 것은 어떤 문제점이 있을까요?

laco-dev reviewed (approved):
피드백 반영하느라 고생 많으셨습니다.

여러가지 다양한 개념 적용해보시는 과정이 보기 좋았습니다.
구조를 미리 나눠두었기 때문에 다음 미션에서는 큰 변경이 없을 것 같네요.

남겨드린 추가적인 피드백을 포함해서, 도전해보고 싶은 것이 있다면 해보셔도 좋습니다.
